addons/hr_holidays/models/res_company.py

```diff
@@ -0,0 +1,135 @@
+        holiday_start_utc = company_tz.localize(datetime.combine(start_date, time.min)).astimezone(pytz.utc).replace(tzinfo=None)
+        holiday_end_utc = company_tz.localize(datetime.combine(end_date, time.max)).astimezone(pytz.utc).replace(tzinfo=None)
+        return holiday_start_utc, holiday_end_utc
+
+    def _generate_public_holidays(self, year_range, convert_datetime=True):
+        response = []
+        subdivision_per_country_dict = holidays.list_supported_countries(include_aliases=True)
+        lang_per_country = holidays.list_localized_countries(include_aliases=True)
+        existing_holidays = self.env["resource.calendar.leaves"].search_read(
+            domain=[
+                ('company_id', 'in', self.ids),
+                ('resource_id', '=', False),
+            ],
+            fields=['company_id', 'date_from', 'date_to'],
+        )
+        existing_holidays_dict = defaultdict(list)
+        for holiday in existing_holidays:
+            company_id = holiday['company_id'][0]
+            existing_holidays_dict[company_id].append((holiday['date_from'], holiday['date_to']))
+
+        for company in self:
+            if not company.country_code:
+                response.append({
+                    'title': self.env._('No Country Code'),
+                    'type': 'danger',
+                    'message': self.env._('Please select a country in %(company)s to load public holidays.', company=company.name),
+                })
+                continue
+            if company.country_code not in subdivision_per_country_dict:
+                response.append({
+                    'title': self.env._('No Public Holidays'),
+                    'type': 'danger',
+                    'message': self.env._('Public holidays are not available for %(country)s country.', country=company.country_id.name),
+                })
+                continue
+
+            company_subdiv = company.state_id.code
+            subdiv = subdivision_per_country_dict.get(company.country_code, None)
+            user_lang_iso_code = self.env["res.lang"]._lang_get(self.env.user.lang).iso_code[:2]
+            if not user_lang_iso_code or user_lang_iso_code not in lang_per_country.get(company.country_code, []):
+                user_lang_iso_code = 'en_US'
+
+            public_holiday_dict = holidays.country_holidays(
+                company.country_code,
+                subdiv=company_subdiv if subdiv and company_subdiv in subdiv else None,
+                years=year_range,
+                language=user_lang_iso_code,
+            )
+
+            overlapped_holidays = False
+            company_tz = pytz.timezone(company.resource_calendar_id.tz)
+            public_holidays_values_dict = {}
+
+            for holiday_date, holiday_name in public_holiday_dict.items():
+                holiday_start_utc, holiday_end_utc = self._convert_to_utc(holiday_date, holiday_date, company_tz)
```

Comment:
Actually the method already exist as `_convert_timezone` in `resource.calendar.leaves`, remove the method and make it method generic in `/odoo/tools/date_utils.py` and call that instead, that way there's no need for an extra method.
If possible, make the change in a separate commit that comes first then do this in your commit relative to public holidays:
```suggestion
                holiday_start_utc = convert_timezone(datetime.combine(holiday_date, time.min), company_tz, pytz.utc)
                holiday_end_utc = convert_timezone(datetime.combine(holiday_date, time.max), company_tz, pytz.utc)
```